### ISSUE TITLE
axios auth does not seem to be propagated on redirect in 1.15.4+

### ISSUE THREAD

**Issue by kevkcc** [follow-redirects/follow-redirects#243]
I am seeing this issue as described in https://github.com/axios/axios/issues/6179

This looks to be regression in 1.15.4 as it both workflows below work in 1.15.3. Whereas in 1.15.4+,

Basically this would return 401 after the redirect as the auth options don't seem to be propagated.
```
await axios.get('url_that_returns_301_and_then_redirects',
        { 
            auth: { 
                username: 'username',
                password: 'password'
            }
        });
```

whereas this works and returns 200:
```
await axios.get('url_that_redirects_and_returns_301',
        headers: {
                'Authorization': 'Basic base64_encoding_of_username_password',
            },
```

Through debugging, I can confirm it gets lost because the below works:
```
await axios.get('url_that_redirects_and_returns_301',
        { 
            auth: { 
                username: 'username',
                password: 'password'
            },
            beforeRedirect: (opts, res) => {
                 opts.auth = 'base64_encoding_of_username_password'
            }
        });
```

**Comment by RubenVerborgh** [comment]
That shouldn't be the case indeed!

Are you sure it's not an HTTP/HTTPS redirect issue as in https://github.com/follow-redirects/follow-redirects/issues/192?

**Comment by kevkcc** [comment]
Yea it shouldn't be an HTTP/HTTPS redirect issue. Both URLs for my case are HTTPS. And even so, it looks like you fixed the issue in 1.14.9 and I'm seeing a broken workflow between 1.15.3 and 1.15.4.

**Comment by ThibaudCrespin** [comment]
Hi ! Any updates on this issue ? I'm facing the same problem with both URLs in HTTPS...

**Comment by RubenVerborgh** [comment]
It's on the backlog; current funds are too limited to put a dedicated person on this. I'm trying on a best-effort basis.

**Comment by GabrielScotaEE** [comment]
I was getting the same problem with version 1.15.2 (also tryed 1.15.3 and 1.15.6) on a login request where the POST request to make the login is a redirectable request, returnin 302 status (if I disable redirects). But if the redirects are enable I'm getting a 200 responso but the login was not success.
After some time searching for what's happening, I've found that in the redirect, the session cookies was not being set, so I got a redirect to login page again. To fix it I had to disable redirects, and check for the response status, if it's on the 300 range I'm getting the response cookies, adding them on headers and finally making the redirect request.

**Comment by kevkcc** [comment]
Yes, if you manually pass in the headers, it will work as I had originally reported (tapping into axios' beforeRedirect handler). Nevertheless, that is only a workaround and this is still a clear regression happening in 1.15.4+. As 1.15.4+ fixes vulnerability [CVE-2023-26159](https://nvd.nist.gov/vuln/detail/CVE-2023-26159), it would be great if this could be fixed and released such that we get a version of this library with both the vulnerability fix and no serious regression.

**Comment by dougkeen** [comment]
I did some research on this issue today, and it looks like it only affects Node v16 and lower. In Node 16, `useNativeURL == true`, and in higher versions, `useNativeURL == false`, and that seems to make the difference. The native `URL` object doesn't carry an `auth` property (only separate `username` and `password`), which seems to be what's necessary to propagate authentication to the redirect URL.

If I were to hazard a guess on how to fix this, there are couple things places that look suspicious:
* [This line](https://github.com/follow-redirects/follow-redirects/blob/35a517c5861d79dc8bff7db8626013d20b711b06/index.js#L456) calls `spreadUrlObject` but doesn't assign it to anything, making it effectively a no-op. Maybe it should be `this._options = spreadUrlObject(redirectUrl, this._options);`?
* We might be able to make `spreadUrlObject` handle native node 16 URL objects by detecting when `username` and `password` are present, but `auth` isn't. Maybe something like:
```javascript
  if (urlObject.username && urlObject.password && !urlObject.auth) {
    spread.auth = `${urlObject.username}:${urlObject.password}`
  }
  ```

**Comment by dougkeen** [comment]
Yup, just confirmed that the `http` protocol (and `https`) that this library wraps only looks at the `auth` field on the `options` object to compute an Authorization header, and disregards `username` and `password` fields.
https://nodejs.org/api/http.html#httprequesturl-options-callback

**Comment by RubenVerborgh** [comment]
> * [This line](https://github.com/follow-redirects/follow-redirects/blob/35a517c5861d79dc8bff7db8626013d20b711b06/index.js#L456) calls `spreadUrlObject` but doesn't assign it to anything, making it effectively a no-op.

It assigns it to the second parameter: https://github.com/follow-redirects/follow-redirects/blob/35a517c5861d79dc8bff7db8626013d20b711b06/index.js#L585

> * We might be able to make spreadUrlObject handle native node 16 URL objects by detecting when username and password are present, but auth isn't. 

Agreed; could you PR? Could also check for `useNativeURL` to only change it in that case.

**Comment by dougkeen** [comment]
> It assigns it to the second parameter:
> 
> https://github.com/follow-redirects/follow-redirects/blob/35a517c5861d79dc8bff7db8626013d20b711b06/index.js#L585
> 
Oh tricky, didn't see the sneaky mutation via object ref.

I can take a crack at a PR...

**Comment by RubenVerborgh** [comment]
Do check https://github.com/follow-redirects/follow-redirects/issues/261 as well.

**Comment by xpan0616** [comment]
Is it fixed ?
